File: Data Structures & Algorithms/is-anagram/submission-0.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        
        if(s.size() != t.size())
        return false;
        int n = s.size();
        vector<int> freq1(26, 0), freq2(26, 0);
        for(int i = 0; i <  n; i++){
        freq1[s[i] - 'a']++;
        freq2[t[i] - 'a']++; 
        }
        for(int i = 0; i < 26; i++){
             if(freq1[i] != freq2[i] )
             return false; 
        }
      return true;
    }
};

File: Data Structures & Algorithms/permutation-string/submission-0.java
class Solution {
    public static boolean isValid(Map<Character, Integer> map1, Map<Character, Integer> map2) {
      if(map1.size() > map2.size()){
        return false;
      }

    for (Map.Entry<Character, Integer> entry : map1.entrySet()) {
        Character key = entry.getKey();
        Integer val1 = entry.getValue();

        if (!map2.containsKey(key)) {
            return false;
        }

        Integer val2 = map2.get(key);
        if (val1 > val2) {
            return false;
        }
    }
    return true;
}
    public boolean checkInclusion(String s1, String s2) {
        int n = s1.length() , m = s2.length();
        HashMap<Character, Integer> freq1 = new HashMap<>();
        HashMap<Character, Integer> freq2 = new HashMap<>();

        for(char c : s1.toCharArray()){
            freq1.merge(c, 1, Integer::sum);
        }
        int l = 0;
        for(int r = 0 ; r < m; r++){
            freq2.merge(s2.charAt(r), 1 , Integer::sum);
            while(r-l+1 > n){
                freq2.merge(s2.charAt(l), -1, Integer::sum);
                if(freq2.get(s2.charAt(l)) == 0){
                    freq2.remove(s2.charAt(l));
                }
                l++;
            }
            if(isValid(freq1, freq2)){
                return true;
            }
        }
        return false;

    }
}
